### PR TITLE
JP-3562: Shower flagging enhancement

### PR DESCRIPTION
**Pull request by mwregan2**
Resolves [JP-3562](https://jira.stsci.edu/browse/JP-3562)

Closes spacetelescope/jwst#8333

This PR addresses several effects that lead to inaccurate flagging of showers. These change improve the estimate of the median difference by using all the integrations. It also masks reference pixels to avoid them being used in the convolution used to detect faint emission.

**Checklist**

- [X] added entry in `CHANGES.rst` (either in `Bug Fixes` or `Changes to API`)
- [x] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [x] added relevant label(s)

### CODE REVIEW

**Review by kmacdonald-stsci — APPROVED** [review]

**Comment by hbushouse** [comment]
@mwregan2 The description above refers to JP-3560, but that Jira ticket concerns updates to the charge_migration step. Is there another JP ticket that this should refer to?

File: src/stcal/jump/jump.py [hunk 1]
```diff
@@ -343,9 +347,9 @@ def detect_jumps(
             slices.insert(
                 i,
                 (
-                    data[:, :, i * yinc : (i + 1) * yinc, :],
-                    gdq[:, :, i * yinc : (i + 1) * yinc, :].copy(),
```
Comment (hbushouse) [inline src/stcal/jump/jump.py:347]: Is it OK to remove the .copy() here? I assume it's not needed?

File: src/stcal/jump/jump.py [hunk 2]
```diff
@@ -864,19 +870,23 @@ def near_edge(jump, low_threshold, high_threshold):
...
     nframes,
     minimum_sigclip_groups,
     snr_threshold=1.3,
     min_shower_area=40,
     inner=1,
     outer=2,
+    donotuse_flag = 1,
     sat_flag=2,
     jump_flag=4,
+    refpix_flag=2147483648,
```
Comment (hbushouse) [inline src/stcal/jump/jump.py:885]: In general it's bad form and we discourage the use of explicit/hardwired DQ flag values in the code. Should import the dqflags definitions and use those instead.

**Review by hbushouse — COMMENTED** [review]
Needs a change log entry, in addition to specific code comments/questions.

**Comment by mwregan2** [comment]
Howard,
  It should be JP-3562
Sorry,
Mike

From: Howard Bushouse ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Thursday, March 14, 2024 at 1:12 PM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] Shower flagging enhancement (PR #248)


@mwregan2<https://urldefense.com/v3/__https:/github.com/mwregan2__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!0klFnl0lH0AfMInlyZIf9LIVL1EnraDg4-ZbAu18eP7GaLZFk9JAnX0b_HGVwlO7z9zlemuL8oDJ-lv-3Ey59hE$> The description above refers to JP-3560, but that Jira ticket concerns updates to the charge_migration step. Is there another JP ticket that this should refer to?

File: src/stcal/jump/jump.py (hunk 1 above)
Comment (mwregan2) [reply src/stcal/jump/jump.py:347]: I would keep it in to prevent corruption of the original gdq.

File: src/stcal/jump/jump.py (hunk 2 above)
Comment (mwregan2) [reply src/stcal/jump/jump.py]: I removed the hard coded DQ values for reference pixels.

File: src/stcal/jump/jump.py [hunk 3]
```diff
@@ -369,9 +371,9 @@ def detect_jumps(
         slices.insert(
             n_slices - 1,
             (
-                data[:, :, (n_slices - 1) * yinc : n_rows, :],
-                gdq[:, :, (n_slices - 1) * yinc : n_rows, :].copy() ,
-                readnoise_2d[(n_slices - 1) * yinc : n_rows, :],
+                data[:, :, (n_slices - 1) * yinc: n_rows, :],
+                gdq[:, :, (n_slices - 1) * yinc: n_rows, :],
```
[the remark this answers is not in the document]
Comment (mwregan2) [reply src/stcal/jump/jump.py:375]: I added it back in.

File: src/stcal/jump/jump.py (hunk 1 above)
Comment (mwregan2) [reply src/stcal/jump/jump.py:347]: I added the copy back

**Comment by mwregan2** [comment]
I think I've addressed all the comments.

**Comment by hbushouse** [comment]
@mwregan2 Several unit tests are failing in the CI test suite. They need fixing.

**Review by hbushouse — APPROVED** [review]
CI tests are all passing now. Code updates look OK to me.

**Pull request merged**